### ISSUE TITLE
Expired and self-signed TLS certificates are accepted if a request to the same IP address worked before

### ISSUE THREAD

**Issue by imgx64** [reactor/reactor-netty#236]
If a request is done to a valid HTTPS URL with proper certificates, then a request to the same IP address but a different domain name and a different (invalid) certificate, the invalid certificates are used.

See this code:
```
HttpClient.create().get("https://badssl.com/").block(); // Without this line, the below requests fail

HttpClient.create().get("https://expired.badssl.com/").block();
HttpClient.create().get("https://self-signed.badssl.com/").block();
HttpClient.create().get("https://untrusted-root.badssl.com/").block();

System.err.println("Reached here. It shouldn't.");
```

### Expected behavior
Exception thrown.

### Actual behavior
`Reached here. It shouldn't.` is printed

### Steps to reproduce
Run the above code.

### Reactor Netty version
0.7.2.RELEASE

### JVM version (e.g. `java -version`)
1.8.0_152

### OS version (e.g. `uname -a`)
Windows 10 version 1703

**Comment by smaldini** [comment]
By default the HttpClient is indeed supporting all ssl but should be / could be restricted.

**Comment by bclozel** [comment]
"Connection coalescing"/"unsharding"/"connection reuse" (or whatever you want to call it) is not strictly forbidden, see [rfc7540](http://httpwg.org/specs/rfc7540.html#reuse).

In this case, the client connection pool cache could:

1. use the "full domain + host IP" as a key to cache connections and not do any unsharding
2. possibly associates a cached connection to several domains if the remote certificate allows multiple domains (like `"*.example.org"`)

But even 2) could be quite challenging if `"first.example.org"` resolves to `192.168.0.1,192.168.0.2` and `"second.example.org"` resolves to `192.168.0.2,192.168.0.3`. You must make sure to only reuse a connection:

* if the IP is resolved in common (here, `192.168.0.2`)
* and if TLS is involved, the server certificate covers both domains

**Comment by violetagg** [comment]
Fixed as part of 0.7.4.RELEASE
